File: 3111.py
import sys

p = None  # pattern
t = None  # text
left = []
right = []

p_len = None  # length of pattern
t_len = None  # length of text

l_idx = None  # index from left
r_idx = None  # index from right


def match_left():
    global left, right, p_len, t_len, l_idx, r_idx
    p_matched_cnt = 0
    while l_idx <= r_idx:
        if p_matched_cnt == p_len:
            l_idx += p_len
            break
        elif (l_idx + p_matched_cnt <= r_idx
                and t[l_idx + p_matched_cnt] == p[p_matched_cnt]):
            p_matched_cnt += 1
        elif p_matched_cnt != 0:
            for i in range(p_matched_cnt):
                left.append(t[l_idx + i])
            l_idx += p_matched_cnt
            p_matched_cnt = 0
        else:
            left.append(t[l_idx])
            l_idx += 1
    if p_matched_cnt == p_len:
        return True
    else:
        return False


def match_right():
    global left, right, p_len, t_len, l_idx, r_idx
    p_matched_cnt = 0
    while r_idx >= l_idx:
        if p_matched_cnt == p_len:
            r_idx -= p_len
            break
        elif (r_idx - p_matched_cnt >= l_idx
                and t[r_idx - p_matched_cnt] == p[(p_len - 1) - p_matched_cnt]):
            p_matched_cnt += 1
        elif p_matched_cnt != 0:
            for i in range(p_matched_cnt):
                right.append(t[r_idx - i])
            r_idx -= p_matched_cnt
            p_matched_cnt = 0
        else:
            right.append(t[r_idx])
            r_idx -= 1
    if p_matched_cnt == p_len:
        return True
    else:
        return False


if __name__ == "__main__":
    p = sys.stdin.readline().rstrip()
    t = sys.stdin.readline().rstrip()
    p_len = len(p)
    t_len = len(t)
    l_idx = 0
    r_idx = t_len - 1
    while True:
        if match_left() == False:
            break
        if match_right() == False:
            break
    result = ""
    for x in left:
        result += x
    right.reverse()
    for x in right:
        result += x
    sys.stdout.write(result)

File: code/3111.cpp
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;


#define MAX_PATTERN_LEN 25
#define MAX_TEXT_LEN 300000


char pattern[MAX_PATTERN_LEN + 1];
char text[MAX_TEXT_LEN + 1];
char result[MAX_TEXT_LEN + 1];
int pLen, tLen;  // pLen: length of pattern, tLen: length of text
int lIdx, rIdx;  // lIdx: index from left side, rIdx: index from right side
vector<char> leftVec;
vector<char> rightVec;


bool matchLeft()
{
  int pMatchedCnt = 0;
  while (lIdx <= rIdx)
  {
    if (pMatchedCnt == pLen)
    {
      lIdx += pLen;
      break;
    }
    else if (lIdx + pMatchedCnt <= rIdx
        && text[lIdx + pMatchedCnt] == pattern[pMatchedCnt])
    {
      pMatchedCnt++;
    }
    else if (pMatchedCnt != 0)
    {
      for (int i = 0; i < pMatchedCnt; i++)
        leftVec.push_back(text[lIdx + i]);
      lIdx += pMatchedCnt;
      pMatchedCnt = 0;
    }
    else
    {
      leftVec.push_back(text[lIdx++]);
    }
  }
  return (pMatchedCnt == pLen);
}

bool matchRight()
{
  int pMatchedCnt = 0;
  while (rIdx >= lIdx)
  {
    if (pMatchedCnt == pLen)
    {
      rIdx -= pLen;
      break;
    }
    else if (rIdx - pMatchedCnt >= lIdx
        && text[rIdx - pMatchedCnt] == pattern[(pLen - 1) - pMatchedCnt])
    {
      pMatchedCnt++;
    }
    else if (pMatchedCnt != 0)
    {
      for (int i = 0; i < pMatchedCnt; i++)
        rightVec.push_back(text[rIdx - i]);
      rIdx -= pMatchedCnt;
      pMatchedCnt = 0;
    }
    else
    {
      rightVec.push_back(text[rIdx--]);
    }
  }
  return (pMatchedCnt == pLen);
}


int main()
{
  scanf("%s%s", pattern, text);
  pLen = strlen(pattern);
  tLen = strlen(text);
  lIdx = 0;
  rIdx = tLen - 1;
  while (true)
  {
    if (matchLeft() == false) break;
    if (matchRight() == false) break;
  }

  int idx = 0;
  for (int i = 0; i < leftVec.size(); i++)
    result[idx++] = leftVec[i];
    //cout << leftVec[i];
  for (int i = rightVec.size() - 1; i >= 0; i--)
    result[idx++] = rightVec[i];
    //cout << rightVec[i];
  idx = 0;
  while (result[idx])
    putchar(result[idx++]);
  putchar('\n');

  return 0;
}
